src/broadcast/broadcast.rs

```diff
@@ -362,44 +362,55 @@ impl<N: NodeIdT> Broadcast<N> {
         }
         // Upon receiving 2f + 1 matching Ready(h) messages, send full
         // `Echo` message to every node who hasn't sent us a `CanDecode`
-        let our_value = self.echos.get(self.our_id()).unwrap();
         if self.count_readys(hash) == 2 * self.netinfo.num_faulty() + 1 {
             // `send_echo` function should take care of which nodes to send the message
-            step.extend(self.send_echo(our_value)?);
+            step.extend(self.send_echo_remaining()?);
         }
 
         Ok(step.join(self.compute_output(hash)?))
     }
 
-    /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
-    fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-       if !self.netinfo.is_validator() {
+    /// Sends `Echo` message to all left nodes.
+    fn send_echo_left(&mut self) -> Result<Step<N>> {
+        if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
+        // Caller should make sure message is inserted. unwrap should never panic.
+        let p = self.echos.get(self.our_id()).unwrap().clone();
```

Comment:
But at this point we haven't inserted it yet! That only happens below (`self.handle_echo(our_id, p)`).
I actually think `send_echo_left` needs to keep its `p` argument.

Reply:
I had inserted the message in the `handle_value` function before calling `send_echo_left` . 
```
// Insert the value received into echos map before sending `Echo` and `EchoHash` to peers.
// Will get warning in `handle_echo` call for our_id.
self.echos.insert(sender_id.clone(), (p.root_hash().clone(), Some(p.clone())));

// Send the proof in an `Echo` message to left nodes
// and `EchoHash` message to right nodes and handle the response.
let echo_steps = self.send_echo_left()?;
let echo_hash_steps = self.send_echo_hash(p.root_hash())?;
Ok(echo_steps.join(echo_hash_steps))
```

But yeah, passing the `p` argument to `send_echo_left` probably makes more sense.